java/src/main/java/com/logicalclocks/hsfs/engine/HudiEngine.java

```diff
@@ -0,0 +1,293 @@
+    //properties.setProperty(DataSourceWriteOptions.PARTITIONPATH_FIELD_OPT_KEY(), partitionColumns.mkString(","));
+    //ComplexKeyGenerator compositeKeyGenerator = new ComplexKeyGenerator(properties);
+    //JavaRDD<HoodieKey> deleteRdd2 = deleteDF.toJavaRDD().map(row ->
+    //    new HoodieKey(compositeKeyGenerator.getRecordKey(row), compositeKeyGenerator.getPartitionPath(row)));
+
+    JavaRDD<HoodieKey> deleteRdd = deleteDF.toJavaRDD().map(row ->
+        new HoodieKey(row.<String>getAs(primaryColumns.head()), row.<String>getAs(partitionColumns.head())));
+
+    FileSystem hopsfsConf = FileSystem.get(sparkSession.sparkContext().hadoopConfiguration());
+    HoodieTimeline commitTimeline = HoodieDataSourceHelpers.allCompletedCommitsCompactions(hopsfsConf,
+        featureGroup.getLocation());
+
+    commitTimeline.getInstants().map(commit -> client.delete(deleteRdd, commit.getTimestamp()));
+  }
+
+  // TODO (davit): this is WIP
+  private Integer rollback(SparkSession sparkSession, String instantTime, String basePath) {
+    HoodieWriteConfig config = HoodieWriteConfig.newBuilder().withPath(basePath)
+        .withIndexConfig(HoodieIndexConfig.newBuilder().withIndexType(HoodieIndex.IndexType.BLOOM).build()).build();
+    JavaSparkContext jsc = JavaSparkContext.fromSparkContext(sparkSession.sparkContext());
+    HoodieWriteClient client = new HoodieWriteClient(jsc, config);
+    if (client.rollback(instantTime)) {
+      LOGGER.info(String.format("The commit \"%s\" rolled back.", instantTime));
+      return 0;
+    } else {
+      LOGGER.warn(String.format("The commit \"%s\" failed to roll back.", instantTime));
+      return -1;
+    }
+  }
+
+  private Map<String, String> setupHudiWriteOpts(FeatureGroup featureGroup) throws IOException, FeatureStoreException {
+
+    Map<String, String> hudiArgs = new HashMap<String, String>();
+
+    hudiArgs.put(Constants.HUDI_TABLE_STORAGE_TYPE, Constants.HUDI_COPY_ON_WRITE);
+
+    Seq<String> primaryColumns = utils.getPrimaryColumns(featureGroup);
+
+    // primary key is required
+    if (primaryColumns.isEmpty()) {
+      throw new FeatureStoreException("For time travel enabled feature groups You must provide at least 1 primary key");
+    }
+
+    hudiArgs.put(Constants.HUDI_RECORD_KEY, primaryColumns.mkString(","));
+
+    // BUG: https://github.com/apache/hudi/pull/1984
+    // hudiArgs.put(Constants.HUDI_KEY_GENERATOR_OPT_KEY, Constants.HUDI_COMPLEX_KEY_GENERATOR_OPT_VAL);
+
+    Seq<String> partitionColumns = utils.getPartitionColumns(featureGroup);
+
+    if (!partitionColumns.isEmpty()) {
+      hudiArgs.put(Constants.HUDI_PARTITION_FIELD, partitionColumns.mkString(","));
+      // For precombine key take 1st particion key
+      hudiArgs.put(Constants.HUDI_PRECOMBINE_FIELD, partitionColumns.head());
+      hudiArgs.put(Constants.HUDI_HIVE_SYNC_PARTITION_FIELDS, partitionColumns.mkString(","));
+    }
+
+    String tableName = featureGroup.getName() + "_" + featureGroup.getVersion();
```

Comment:
We already have a method in `Util.java` that does the same thing. We should that one instead.